masu/database/provider_status_accessor.py

```diff
@@ -0,0 +1,212 @@
+        self._provider_uuid = provider_uuid
+        self._table = self.get_base().classes.api_providerstatus
+        self._obj = self._get_db_obj_query().first()
+
+        # default to READY if there is no previous status.
+        if self._obj is None:
+            ready_status = {'provider_id': self.provider.id,
+                            'status': ProviderStatusCode.READY,
+                            'last_message': 'none',
+                            'timestamp': datetime.now(timezone.utc),
+                            'retries': 0}
+            self.add(**ready_status)
+            self.commit()
+            self._obj = self._get_db_obj_query().first()
+
+    # pylint: disable=arguments-differ
+    def _get_db_obj_query(self):
+        """
+        Return the sqlachemy query for the provider object.
+
+        Args:
+            None
+        Returns:
+            (sqlalchemy.orm.query.Query): "SELECT public.api_customer.group_ptr_id ..."
+        """
+        provider_accessor = ProviderDBAccessor(self._provider_uuid)
+        self.provider = provider_accessor.get_provider()
+        if not self.provider:
+            raise MasuProviderError(f'Unknown provider: {self._provider_uuid}')
+        return super()._get_db_obj_query(provider_id=self.provider.id)
+
+    def get_status(self):
+        """
+        Return the provider status.
+
+        Args:
+            None
+        Returns:
+            (int): Status code
+        """
+        return self._obj.status
+
+    def get_provider_uuid(self):
+        """
+        Return the provider uuid.
+
+        Args:
+            None
+        Returns:
+            (uuid.uuid4): Provider uuid
+        """
+        return self._provider_uuid
+
+    def get_last_message(self):
+        """
+        Return the provider last_message.
+
+        Args:
+            None
+        Returns:
+            (str): Error message, if any
+        """
+        return self._obj.last_message
+
+    def get_timestamp(self):
+        """
+        Return the provider status timestamp.
+
+        Args:
+            None
+        Returns:
+            (datetime): Timestamp for the status
+        """
+        return self._obj.timestamp
+
+    def get_retries(self):
+        """
+        Return the provider status retry count.
+
+        Args:
+            None
+        Returns:
+            (int): Status retry count
+        """
+        return self._obj.retries
+
+    def is_backing_off(self):
+        """Determine if the provider is waiting to retry."""
+        def backoff(interval, maximum=64):
+            """Exponential back-off."""
+            # (12 + interval) is used to select desirable values from the
+            # exponential progression.
+            return min(maximum, (2 ** (12 + interval))) + (random.randint(0, 1000) / 1000.0)
+
+        retries = self.get_retries()
+        timestamp = self.get_timestamp()
+        backoff_threshold = timestamp + timedelta(seconds=backoff(retries,
```

Comment:
Still thinking on the date time based backoff scheme.  Thinking around how this works with our regular cadence of polling attempts and the irregular nature of AWS report updates.  I'll get back to you :)

Reply:
The reason I went with the time-based mechanism is because it allows the poller to be configured however the admin wants, but no matter how frequently the poller runs, we're enforcing the same backoff window. 

Otherwise, the rate at which we reach error states is a function of the poller frequency. This means that the shorter the poller intervals are, the more likely we see false-positive error states.